src/java/org/apache/cassandra/index/sai/disk/vector/VectorSourceModel.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.disk.vector;
+
+import static org.apache.cassandra.index.sai.disk.vector.VectorCompression.CompressionType.BINARY_QUANTIZATION;
+import static org.apache.cassandra.index.sai.disk.vector.VectorCompression.CompressionType.PRODUCT_QUANTIZATION;
+
+public enum VectorSourceModel
+{
+    ADA002,
+    OPENAI_V3_SMALL,
+    OPENAI_V3_LARGE,
+    BERT,
+    GECKO,
+
+    OTHER;
+
+    public static VectorSourceModel fromString(String value)
+    {
+        return valueOf(value.toUpperCase());
+    }
+
+    public VectorCompression preferredCompression(int originalDimension)
+    {
+        // if model is specified, use specifically tuned compression parameters
+        switch (this)
+        {
+            case ADA002:
+                return new VectorCompression(BINARY_QUANTIZATION, originalDimension / 8);
+            case OPENAI_V3_SMALL:
+                return new VectorCompression(PRODUCT_QUANTIZATION, originalDimension / 16);
+            case OPENAI_V3_LARGE:
+                return new VectorCompression(PRODUCT_QUANTIZATION, originalDimension / 16);
+            case BERT:
+                // VSTODO test if we can be more aggressive here
+                return new VectorCompression(PRODUCT_QUANTIZATION, originalDimension / 4);
+            case GECKO:
+                return new VectorCompression(PRODUCT_QUANTIZATION, originalDimension / 4);
+        }
+
+        // Model is unspecified / unknown, so we guess.
+        // Guessing heuristic is simple: 1536 dimensions is probably ada002, so use BQ.  Otherwise, use PQ.
+        assert this == OTHER : this;
+        if (originalDimension == 1536)
+            return new VectorCompression(BINARY_QUANTIZATION, originalDimension / 8);
```

Comment:
I notice that we never use the `compressToBytes` for `BINARY_QUANTIZATION`. Is that because binary quantization always compresses by a factor of 8, as seen in this method?

Reply:
BQ means we turn the original into a single bit.  Since the original is always f32 then BQ is always 32x for us.